src/scrot.c

```diff
@@ -285,33 +290,53 @@ static Imlib_Image scrotGrabAutoselect(void)
     return im;
 }
 
-/* similar to sleep, but deals with EINTR.
- * ideally, we'd want to sleep against an absolute time to prevent any drift
- * but clock_nanosleep doesn't seem to be widely implemented.
- */
-static void scrotSleep(int sec)
+void scrotDoDelay(void)
 {
-    assert(sec > 0);
-    struct timespec delay = { .tv_sec = sec };
-    while (nanosleep(&delay, &delay) < 0 && errno == EINTR);
+    if (!opt.delay)
+        return;
+    if (opt.countdown) {
+        dprintf(STDERR_FILENO, "Taking shot in ");
+        for (int i = opt.delay; i > 0; i--) {
+            dprintf(STDERR_FILENO, "%d.. ", i);
+            opt.delayStart.tv_sec++;
+            scrotWaitUntil(&opt.delayStart);
+        }
+        dprintf(STDERR_FILENO, "0.\n");
+    } else {
+        opt.delayStart.tv_sec += opt.delay;
+        scrotWaitUntil(&opt.delayStart);
+    }
 }
 
-void scrotDoDelay(void)
+/* scrotWaitUntil: clock_nanosleep with a simpler interface and no EINTR nagging
+ */
+static void scrotWaitUntil(const struct timespec *time)
 {
-    if (opt.delay) {
-        if (opt.countdown) {
-            int i;
-
-            dprintf(STDERR_FILENO, "Taking shot in %d.. ", opt.delay);
-            scrotSleep(1);
-            for (i = opt.delay - 1; i > 0; i--) {
-                dprintf(STDERR_FILENO, "%d.. ", i);
-                scrotSleep(1);
-            }
-            dprintf(STDERR_FILENO, "0.\n");
-        } else
-            scrotSleep(opt.delay);
-    }
+    int savedErrno = errno;
+#if defined(HAVE_CLOCK_NANOSLEEP)
+    while (clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, time, NULL)
```

Comment:
One more thing: `MONOTONIC` doesn't count time when the system is suspended. `CLOCK_BOOTTIME` does, but that seems like it's linux specific and non-POSIX.

Reply:
Interesting, CLOCK_BOOTTIME is also on OpenBSD. Will write a test program and see if it has the same caveat.

Reply:
FreeBSD also has it: https://man.freebsd.org/cgi/man.cgi?query=clock_gettime&manpath=FreeBSD+13.2-RELEASE+and+Ports

Doesn't mention anything about what happens during system suspension though.

> Guess we'll need #ifdef __linux__ to work around that Linux bug.

If BOOTTIME is portable enough then we should just use it.

Reply:
>If BOOTTIME is portable enough then we should just use it.

Looks like netbsd, haiku etc doesn't have it. Perhaps:

```c
#ifdef CLOCK_BOOTTIME
    #define SCROT_CLOCK CLOCK_BOOTTIME
#else
    #define SCROT_CLOCK CLOCK_MONOTONIC
#endif
```

Reply:
I went with the standard interface everywhere except for the known broken platform.

Reply:
Are you sure that not accounting for system suspend is linux specific behavior? Because POSIX doesn't mention anything about system suspend (neither does FreeBSD manpage).

Reply:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/clock_gettime.html says:
```plain
If the Monotonic Clock option is supported, all implementations shall support 
a clock_id of CLOCK_MONOTONIC defined in <time.h>. This clock represents the
monotonic clock for the system. For this clock, the value returned by 
clock_gettime() represents the amount of time (in seconds and nanoseconds) 
since an unspecified point in the past (for example, system start-up time, or 
the Epoch). This point does not change after system start-up time.
```
It says amount of seconds since that point in the past, whether the machine was suspended or not does not change how much time passed. Linux [tried to fix its CLOCK_MONOTONIC](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d6ed449afdb38f89a7b38ec50e367559e1b8f71f) once, but then [reverted the change](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a3ed0e4393d6885b4af7ce84b437dc696490a530) because userspace applications depended on the behavior.